### ISSUE TITLE
Add the SAHI Algorithm

### ISSUE THREAD

**Issue by kadirnar** [kadirnar/segment-anything-video#7]

**Comment by 1kaiser** [comment]
[here](https://colab.research.google.com/github/1kaiser/Snow-cover-area-estimation/blob/main/Serial_MLP_Image_Train_Inference_JAX.ipynb#scrollTo=FUxTs7XpkfIv&line=8&uniqifier=1) SAHI kind of usage

**Comment by kadirnar** [comment]
I will give the boxes output of the SAHI algorithm to the SAM model. The SAM model doesn't work well for small objects.

**Comment by 1kaiser** [comment]
This is a function that generates the coordinates of the four corners of all the possible square boxes that can be extracted from an image of a given size. The function takes as input the height and width of the image and optionally the height and width of the desired slices, as well as the overlap ratios along the height and width dimensions. The output is a list of coordinates for each slice box.

Here is a possible rephrased code for better understanding:

```
def get_slice_bboxes(image_height, image_width, slice_height=224, slice_width=224, overlap_height_ratio=0.0, overlap_width_ratio=0.0):
    slice_bboxes = []
    y_max = y_min = 0
    y_overlap, x_overlap = (int(overlap_height_ratio * slice_height), int(overlap_width_ratio * slice_width)) if slice_height and slice_width else (None, None)
    
    while y_max < image_height:
        x_min = x_max = 0
        y_max = y_min + slice_height
        
        while x_max < image_width:
            x_max = x_min + slice_width
            
            if y_max > image_height or x_max > image_width:
                xmax, ymax, xmin, ymin = (min(image_width, x_max), min(image_height, y_max), max(0, xmax - slice_width), max(0, ymax - slice_height))
                slice_bboxes.append([xmin, ymin, xmax, ymax])
            else:
                slice_bboxes.append([x_min, y_min, x_max, y_max])
                
            x_min = x_max - x_overlap if x_overlap is not None else x_max
        
        y_min = y_max - y_overlap if y_overlap is not None else y_max
    
    return slice_bboxes

```

This function takes in two inputs: i and b. The variable i represents the slice box of an image, while b represents the input image(s). The function then uses the slice box to extract the corresponding parts of the input images, slice them into smaller sub-images, and return the resulting sub-images.

The function first extracts the first element of b (i.e., b[0]) and swaps the axes to convert the shape of the input images from (batch_size, height, width, channel) to (batch_size, channel, height, width). The extracted input images are then sliced into smaller sub-images using the slice box i. The resulting sub-images are concatenated along the batch axis and converted to the appropriate shape (i.e., (batch_size, height, width, channel)).

The same process is repeated for the second element of b (i.e., b[1]), which is expected to represent the target images. The resulting sub-images are returned as JAX arrays. Additionally, the function prints the shape of the resulting sub-images for both the input and target images.

```
def slicedImageInput(i, b):
    b1, b2 = b[0].swapaxes(1,3).swapaxes(2,3), b[1].swapaxes(1,3).swapaxes(2,3)
    ximages, yimages = [], []
    for j in range(b1.shape[3]):
        a = b1[:, i[1]:i[3], i[0]:i[2], j] if j == 0 else np.append(a, b1[:, i[1]:i[3], i[0]:i[2], j], axis=0)
    ximages = np.expand_dims(a, axis=0).swapaxes(1,2).swapaxes(2,3)
    for j in range(b2.shape[3]):
        a = b2[:, i[1]:i[3], i[0]:i[2], j] if j == 0 else np.append(a, b2[:, i[1]:i[3], i[0]:i[2], j], axis=0)
    yimages = np.expand_dims(a, axis=0).swapaxes(1,2).swapaxes(2,3)
    print(jnp.asarray(ximages).shape, jnp.asarray(yimages).shape)
    return jnp.asarray(ximages), jnp.asarray(yimages)

```

**Comment by kadirnar** [comment]
You are great. Have you tested the SAM models? 

https://github.com/obss/sahi/blob/26acb891168a6338f2b9c89b2f002ded968d6cb5/sahi/prediction.py#L158

I will code using SAHI library. I want to use object detection model + SAHI + SAM. But do you want to add SAHI to the SAM model? Or do you want to do this with the jax library? If you send a Pull Request, I will accept it.

**Comment by kadirnar** [comment]
I added manual box feature. You can try this.

# For manuel box selection
from metaseg import SegManualMaskGenerator

seg_manual_mask_generator = SegManualMaskGenerator().save_image(
    source="image.jpg",
    model_type="vit_l", # vit_l, vit_h, vit_b
    x0=100,
    y0=100,
    x1=200,
    y1=200,
)
